code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/builders/AepPushNotificationBuilder.kt

```diff
@@ -0,0 +1,971 @@
+        } catch (exception: IllegalArgumentException) {
+            Log.trace(
+                PushTemplateConstants.LOG_TAG,
+                SELF_TAG,
+                "Unrecognized hex string passed to Color.parseColor(), custom color will not be applied to the notification icon."
+            )
+        }
+    }
+
+    /**
+     * Returns the resource id for the drawable with the given name. The file must be in the
+     * res/drawable directory. If the drawable file is not found, 0 is returned.
+     *
+     * @param iconName the name of the icon file
+     * @param context the application [Context]
+     * @return the resource id for the icon with the given name
+     */
+    private fun getIconWithResourceName(
+        iconName: String?,
+        context: Context
+    ): Int {
+        return if (iconName.isNullOrEmpty()) {
+            0
+        } else context.resources.getIdentifier(iconName, "drawable", context.packageName)
+    }
+
+    /**
+     * Checks if the icon is valid.
+     *
+     * @param icon the icon to be checked
+     * @return true if the icon is valid, false otherwise
+     */
+    private fun isValidIcon(icon: Int): Boolean {
+        return icon > 0
+    }
+
+    /**
+     * Creates a notification channel if the device is running on Android O or higher. If the channel
+     * already exists, the same channel is used. A default channel ID will be used if no channel ID
+     * is received from the payload.
+     *
+     * @param context the application [Context]
+     * @param channelId [String] containing the notification channel id
+     * @param customSound `String` containing the custom sound to apply on the notification channel
+     * @param importance [Int] containing the notification importance
+     * @return A [String] containing the created or existing channel ID
+     */
+    internal fun createChannel(
+        context: Context,
+        channelId: String?,
+        customSound: String?,
+        importance: Int
+    ): String {
+        // No channel creation required.
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            return channelId ?: PushTemplateConstants.DEFAULT_CHANNEL_ID
+        }
+
+        // Decide on which channel to use provided or default
+        val notificationManager =
+            context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+        val channelIdToUse = channelId ?: PushTemplateConstants.DEFAULT_CHANNEL_ID
+
+        // create the channel
+        createNotificationChannel(
+            context,
+            notificationManager,
+            channelIdToUse,
+            customSound,
+            importance
+        )
+        // return the id of the channel.
+        return channelIdToUse
+    }
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    private fun createNotificationChannel(
+        context: Context,
+        notificationManager: NotificationManager,
```

Comment:
Don't think we need to accept this as a parameter. Since we already have the context we can do 
`val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager`

Reply:
If a previous call to this method already has/needs the system service always, I prefer having this as a parameter rather than re-fetching.

Reply:
It is currently a private method called only by `createChannel` which always refetches it. We can move it out to the builder class if it makes more sense?

Reply:
Will address in subsequent PR when moving it to extensions